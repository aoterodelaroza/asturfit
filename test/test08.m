#! /usr/bin/octave -q
% Copyright (C) 2010 Victor Lua~na and Alberto Otero-de-la-Roza
%
% This octave routine is free software: you can redistribute it and/or modify
% it under the terms of the GNU General Public License as published by
% the Free Software Foundation, either version 3 of the License, or (at
% your option) any later version. See <http://www.gnu.org/licenses/>.
%
% The routine distributed in the hope that it will be useful, but WITHOUT
% ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
% FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
% more details.

% Test08: Bootstrap fit to a data set with noise (disperse outlier points)

addpath("../src/");
global nelectrons
hybohr3togpa = 2*14710.50498740275538944426;

# read in a smooth data set
[v,e] = readEVdata('mgo-sety1.dat',1);
vref = median(v);

# fit to the smooth set
[c,s] = avgstrainfit(v,e,vref,10,:,:,1);

# now we add artificially some noise, only to the energy
noutliers = 10;
[vn,en] = noisify(v, e, noutliers, 0, 0.004, 1);

# lets use the bootstrap sampling and see if all outliers are detected
sample = 1000;
chance = (1-(1-0.5^noutliers)^sample);
printf("\n");
printf("Number of outliers (artificially added): %d\n", noutliers);
printf("Bootstrap sample size: %d\n", sample);
printf("Chance of finding all outliers is %.6e\n", chance);
[cb,sb] = strainbootstrap(vn, en, vref, :, sample, :, 1);

bad = sb.outliers;
good = 1:length(vn); good(bad) = [];
fileplot = 'test08-ev.eps';
plot(vn,sb.Efit,'-r;BS fit;', vn(good),en(good),'ob;good data;'...
    , vn(bad),en(bad),'om;bad data;', 'markersize', 12);
xlabel('V (bohr^3)'); ylabel('E (hartree)'); grid('on');
print(fileplot, '-FHelvetica:24', '-depsc');
printf('See the E(V) curve in the file: %s\n', fileplot);
