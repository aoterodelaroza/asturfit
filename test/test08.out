warning: mark_as_command is obsolete and will be removed from a future version of Octave
readEVdata: Reading input file -> mgo-sety1.dat
Input volume unit: bohr3
Input energy unit: ry
* Energy and volume ranges 
  V-range (bohr^3) : [71.5195563798 , 143.0391127595] dV = 71.5195563796
  E-range (hartree) : [-17.1721409750 , -17.0038769900] dE = 0.1682639850
  Number of data points: 129
  Number of electrons: 20



avgpolyfit: AVERAGE OF eulerian STRAIN POLYNOMIALS
Volume reference (V0): 107.279335
Range of degrees: 2--10
Number of polynomials: 9

Properties at the minimum of each polynomial:
--i- npol data npar --SSerr-- ---w---- ----Vmin--- ----Emin--- ----Bmin--- ---B1min--- ---B2min---
   1    4  129    5  5.11e-11 0.146644  124.824303  -17.172142  169.303010    4.160528 -0.024669721 0.001192786
   2    5  129    6  4.96e-11 0.145614  124.823843  -17.172142  169.308984    4.161944 -0.024887896 0.001212891
   3    6  129    7  4.65e-11 0.144928  124.823413  -17.172142  169.334344    4.162531 -0.025427966 0.001295810
   4    7  129    8  4.65e-11 0.143805  124.823407  -17.172142  169.333353    4.162637 -0.025412950 0.001290615
   5    8  129    9  4.65e-11 0.142693  124.823471  -17.172142  169.334356    4.162043 -0.025389155 0.001305273
   6    9  129   10  4.63e-11 0.141632  124.823722  -17.172142  169.328200    4.159890 -0.024936147 0.001312367
   7    3  129    4  1.91e-10 0.134682  124.821340  -17.172141  169.268283    4.171190 -0.025412395 0.001238681
   8    2  129    3  2.14e-08 0.000003  124.823661  -17.172156  170.041394    4.115715 -0.023629506 0.001146551
   9    1  129    2  2.46e-06 0.000000  124.603461  -17.172260  175.758259    4.000000 -0.022126351 0.001007127

Average properties (weighted polynomials):
------ ---volume-- ---energy-- --B-(GPa)-- ----B1p---- B2p-(1/GPa) B3p--(1/GPa^2)
-mean-  124.823378  -17.172142  169.316150    4.162890   -0.025159    0.001263882
stdvev    0.000857    0.000000    0.022394    0.003411    0.000298    0.000045093

noisify: Added noise to 9 points of 129
--point-- ------xold----- ------xnew----- ------yold----- ------ynew-----
       27       86.046966       86.046966      -17.106804      -17.117639
       45       96.104404       96.104404      -17.142416      -17.204568
       51       99.456883       99.456883      -17.150262      -17.122386
       52      100.015630      100.015630      -17.151413      -17.091270
       63      106.161842      106.161842      -17.161546      -17.125720
       82      116.778026      116.778026      -17.170469      -17.203048
       90      121.247998      121.247998      -17.171833      -17.136773
      123      139.686634      139.686634      -17.167948      -17.178298
      125      140.804127      140.804127      -17.167362      -17.196488

Number of outliers (artificially added): 10
Bootstrap sample size: 1000
Chance of finding all outliers is 6.235762e-01


sb2: BOOTSTRAP ANALYSIS OF eulerian STRAIN POLYNOMIALS
Volume reference (V0): 107.279335
Fitting mode: average of polynomials up to 12 degree
Samples tried, Pol. rejected/accepted:: 1000 0 11000

 Properties at the minimum of each polynomial:
--i- --npol-- data npar --SSerr-- ---w---- ----Vmin--- ----Emin--- ----Bmin--- ---B1min--- ---B2min---
   1    10454   57    5  1.51e-11 0.120173  124.823950  -17.172142  169.306056    4.160638 -0.024681109 0.001193521
   2    10455   57    6  1.47e-11 0.118092  124.823612  -17.172142  169.310651    4.161849 -0.024872803 0.001211446
   3    10456   57    7  1.47e-11 0.115759  124.823542  -17.172142  169.315247    4.161953 -0.024977621 0.001227985
   4    10457   57    8  1.47e-11 0.113449  124.823516  -17.172142  169.312287    4.162339 -0.024932692 0.001210016
   5    10458   57    9  1.44e-11 0.111599  124.823944  -17.172142  169.314164    4.158176 -0.024616551 0.001299576
   6    10459   57   10  1.35e-11 0.110714  124.824829  -17.172142  169.281939    4.150401 -0.022442991 0.001252916
   7    10460   57   11  1.30e-11 0.109440  124.824966  -17.172142  169.242057    4.151467 -0.020206650 0.000790729
   8    10461   57   12  1.27e-11 0.108101  124.825459  -17.172142  169.272700    4.141230 -0.021425135 0.001695233
   9    10453   57    4  6.60e-11 0.092674  124.821624  -17.172141  169.263139    4.171265 -0.025414643 0.001238841
  10    10452   57    3  8.38e-09 0.000000  124.818830  -17.172153  170.106397    4.114620 -0.023612555 0.001144970
  11        1   61    2  6.47e-03 0.000000  128.340838  -17.171483  147.791491    4.000000 -0.026313348 0.001424350
  12        2   61    3  6.47e-03 0.000000  129.246019  -17.171328  133.553243    4.299394 -0.032031561 0.002068717
  13        3   61    4  6.37e-03 0.000000  127.806554  -17.172629  226.890772    0.342136 -0.001882973 0.001028457
  14        4   61    5  6.37e-03 0.000000  127.552133  -17.172468  213.530961    1.653466 -0.031367991 0.001008063
  15        5   61    6  6.15e-03 0.000000  122.845530  -17.173276  377.019558    8.560838 -0.729169043 0.064527171
  16        6   61    7  6.15e-03 0.000000  122.764754  -17.173617  422.778871    7.351037 -0.664703980 0.049560619
  17        7   61    8  5.74e-03 0.000000  117.946370  -17.172565  577.926402   19.648456 -1.224677647 0.120789585
  18        8   61    9  5.68e-03 0.000000  117.758284  -17.173698  821.517662   14.347136 -0.793817523 0.044208929
  19        9   61   10  5.46e-03 0.000000  115.451554  -17.172631  931.852285   21.526002 -0.952588424 0.060261125
  20       10   61   11  5.28e-03 0.000000  114.621412  -17.174413 1680.784542   14.318371 -0.529611896 0.013591666
  21       11   61   12  5.18e-03 0.000000  113.901171  -17.173515 1522.695248   21.153813 -0.678194524 0.024780352
  22       12   64    2  7.52e-03 0.000000  128.059046  -17.170948  148.786340    4.000000 -0.026137405 0.001405366
  23       13   64    3  7.50e-03 0.000000  130.080914  -17.170881  121.326218    4.580252 -0.039610842 0.002993306
  24       14   64    4  7.35e-03 0.000000  127.353121  -17.172419  250.345614   -0.331225 -0.009812100 0.000884600
  25       15   64    5  7.33e-03 0.000000  127.748628  -17.173035  300.171237   -3.067714 -0.014508443 0.001424841

Average properties (weighted polynomials):
------ ---volume-- ---energy-- --B-(GPa)-- ----B1p---- B2p-(1/GPa) B3p--(1/GPa^2)
-mean-  124.823969  -17.172142  169.292017    4.157601   -0.023738    0.001234264
stdvev    0.001000    0.000000    0.025110    0.008009    0.001753    0.000212953

Average properties (equally weighted polynomials):
------ ---volume-- ---energy-- --B-(GPa)-- ----B1p---- B2p-(1/GPa) B3p--(1/GPa^2)
-mean-  129.411707 1185258628.668495 1524310969310773.750000   -3.389148 -9500.995416 660283931.090076089
stdvev    2.736798 124388065853.764206 159907780502563040.000000   67.334085 1053168.763940 69878950012.651596069


Properties of the average polynomial:
Vmin, Emin: 132.250904 -17.171665
SSerr, SStot, R2, 1-R2: 1.290442135e-02 2.677805572e-01 0.951809715056 4.82e-02
Data (min/max/mode/mean/std): 47 83 63 64.5610 5.7265

Detect outlier points (lambda=10.000000):
Ecentral (median): 4.832764e-07
Edisp    (median): 3.444463e-07
-npt- out ---volume-- ---energy-- ---error---
    1   0   71.519556  -17.003877  1.7848e-07
    2   0   72.078303  -17.009504  7.0207e-07
    3   0   72.637049  -17.014963  1.1694e-06
    4   0   73.195796  -17.020260  1.2140e-06
    5   0   73.754543  -17.025400  1.2357e-06
    6   0   74.313289  -17.030386  1.5375e-06
    7   0   74.872036  -17.035225  1.9392e-06
    8   0   75.430782  -17.039923  6.6765e-09
    9   0   75.989529  -17.044480  1.7267e-07
   10   0   76.548275  -17.048902  2.2713e-07
   11   0   77.107022  -17.053194  8.3641e-07
   12   0   77.665768  -17.057358  5.1031e-07
   13   0   78.224515  -17.061398  5.7818e-10
   14   0   78.783261  -17.065318  4.8967e-07
   15   0   79.342008  -17.069123  1.2564e-07
   16   0   79.900754  -17.072813  3.9542e-07
   17   0   80.459501  -17.076395  1.0056e-09
   18   0   81.018247  -17.079869  8.6853e-08
   19   0   81.576994  -17.083239  7.3347e-07
   20   0   82.135741  -17.086510  5.5625e-08
   21   0   82.694487  -17.089682  8.0636e-08
   22   0   83.253234  -17.092758  4.8957e-07
   23   0   83.811980  -17.095744  9.2963e-07
   24   0   84.370727  -17.098638  4.7903e-07
   25   0   84.929473  -17.101444  5.1272e-07
   26   0   85.488220  -17.104166  1.3883e-07
   27   1   86.046966  -17.117639  1.0835e-02
   28   0   86.605713  -17.109362  6.6108e-07
   29   0   87.164459  -17.111841  1.1710e-06
   30   0   87.723206  -17.114246  4.1660e-07
   31   0   88.281952  -17.116576  2.9926e-07
   32   0   88.840699  -17.118834  6.7193e-07
   33   0   89.399445  -17.121021  6.4773e-07
   34   0   89.958192  -17.123139  1.3271e-08
   35   0   90.516939  -17.125192  8.7066e-07
   36   0   91.075685  -17.127179  3.3093e-07
   37   0   91.634432  -17.129103  4.4331e-07
   38   0   92.193178  -17.130967  9.4305e-07
   39   0   92.751925  -17.132770  4.0426e-07
   40   0   93.310671  -17.134514  4.7709e-07
   41   0   93.869418  -17.136203  1.1892e-06
   42   0   94.428164  -17.137835  6.7606e-08
   43   0   94.986911  -17.139414  5.1221e-07
   44   0   95.545657  -17.140940  4.2934e-07
   45   1   96.104404  -17.204568  6.2153e-02
   46   0   96.663150  -17.143840  1.2570e-07
   47   0   97.221897  -17.145217  1.9218e-07
   48   0   97.780643  -17.146546  3.3422e-07
   49   0   98.339390  -17.147829  9.2636e-07
   50   0   98.898137  -17.149068  3.8555e-07
   51   1   99.456883  -17.122386  2.7877e-02
   52   1  100.015630  -17.091270  6.0144e-02
   53   0  100.574376  -17.152523  1.1035e-06
   54   0  101.133123  -17.153593  9.1406e-07
   55   0  101.691869  -17.154623  3.0923e-07
   56   0  102.250616  -17.155615  5.7597e-07
   57   0  102.809362  -17.156568  1.0856e-07
   58   0  103.368109  -17.157484  4.8328e-07
   59   0  103.926855  -17.158366  4.3895e-07
   60   0  104.485602  -17.159211  4.6716e-08
   61   0  105.044348  -17.160023  4.7149e-07
   62   0  105.603095  -17.160801  1.2835e-07
   63   1  106.161842  -17.125720  3.5826e-02
   64   0  106.720588  -17.162260  4.6211e-07
   65   0  107.279335  -17.162943  6.4121e-07
   66   0  107.838081  -17.163595  9.1153e-07
   67   0  108.396828  -17.164220  6.8143e-07
   68   0  108.955574  -17.164814  2.1620e-07
   69   0  109.514321  -17.165380  5.5192e-08
   70   0  110.073067  -17.165919  6.0283e-07
   71   0  110.631814  -17.166432  9.2559e-07
   72   0  111.190560  -17.166918  2.6923e-07
   73   0  111.749307  -17.167378  3.3137e-07
   74   0  112.308053  -17.167814  1.0177e-08
   75   0  112.866800  -17.168225  1.8930e-07
   76   0  113.425546  -17.168612  8.1996e-08
   77   0  113.984293  -17.168976  2.1461e-07
   78   0  114.543040  -17.169318  6.3255e-09
   79   0  115.101786  -17.169638  1.6899e-07
   80   0  115.660533  -17.169936  8.4124e-08
   81   0  116.219279  -17.170212  5.2254e-07
   82   1  116.778026  -17.203048  3.2579e-02
   83   0  117.336772  -17.170705  3.2083e-07
   84   0  117.895519  -17.170921  5.4318e-07
   85   0  118.454265  -17.171118  6.3646e-07
   86   0  119.013012  -17.171296  9.0065e-07
   87   0  119.571758  -17.171458  5.4529e-07
   88   0  120.130505  -17.171601  1.7280e-06
   89   0  120.689251  -17.171726  1.7245e-06
   90   1  121.247998  -17.136773  3.5059e-02
   91   0  121.806744  -17.171924  9.6206e-07
   92   0  122.365491  -17.171998  5.2152e-07
   93   0  122.924238  -17.172057  9.0098e-07
   94   0  123.482984  -17.172101  8.4521e-07
   95   0  124.041731  -17.172128  2.7277e-07
   96   0  124.600477  -17.172141  1.7578e-07
   97   0  125.159224  -17.172139  5.7035e-07
   98   0  125.717970  -17.172124  2.7790e-07
   99   0  126.276717  -17.172094  7.1909e-08
  100   0  126.835463  -17.172051  4.1429e-07
  101   0  127.394210  -17.171994  5.1422e-07
  102   0  127.952956  -17.171926  2.6084e-07
  103   0  128.511703  -17.171844  6.6662e-07
  104   0  129.070449  -17.171750  8.5715e-08
  105   0  129.629196  -17.171643  3.0763e-07
  106   0  130.187942  -17.171526  5.3727e-07
  107   0  130.746689  -17.171396  2.0240e-07
  108   0  131.305436  -17.171255  2.8585e-07
  109   0  131.864182  -17.171103  2.3162e-07
  110   0  132.422929  -17.170940  5.2874e-10
  111   0  132.981675  -17.170767  7.4777e-08
  112   0  133.540422  -17.170584  1.5183e-06
  113   0  134.099168  -17.170390  1.1112e-06
  114   0  134.657915  -17.170187  1.1246e-06
  115   0  135.216661  -17.169973  6.8460e-07
  116   0  135.775408  -17.169751  2.3730e-07
  117   0  136.334154  -17.169519  8.6662e-08
  118   0  136.892901  -17.169278  1.6206e-07
  119   0  137.451647  -17.169028  3.4926e-07
  120   0  138.010394  -17.168770  7.9434e-07
  121   0  138.569140  -17.168504  8.7421e-07
  122   0  139.127887  -17.168230  8.4175e-07
  123   1  139.686634  -17.178298  1.0350e-02
  124   0  140.245380  -17.167659  9.8308e-07
  125   1  140.804127  -17.196488  2.9128e-02
  126   0  141.362873  -17.167056  1.0934e-06
  127   0  141.921620  -17.166744  1.0074e-06
  128   0  142.480366  -17.166423  9.9814e-08
  129   0  143.039113  -17.166096  1.0211e-07

Outliers (9): 27, 45, 51, 52, 63, 82, 90, 123, 125,
See the E(V) curve in the file: test08-ev.eps
